src/dpdkstat.c

```diff
@@ -0,0 +1,809 @@
+  if (reset) {
+    g_configuration->eal_initialized = 0;
+    g_configuration->num_ports = 0;
+    g_configuration->xstats = NULL;
+    g_configuration->num_xstats = 0;
+    for (int i = 0; i < RTE_MAX_ETHPORTS; i++)
+      g_configuration->num_stats_in_port[i] = 0;
+  }
+  close(g_configuration->helper_pipes[1]);
+  int err = kill(g_configuration->helper_pid, SIGKILL);
+  if (err) {
+    char errbuf[ERR_BUF_SIZE];
+    WARNING("dpdkstat: error sending kill to helper: %s", sstrerror(errno, errbuf,
+      sizeof (errbuf)));
+  }
+
+  return 0;
+}
+
+static int dpdk_helper_spawn(enum DPDK_HELPER_ACTION action)
+{
+  char errbuf[ERR_BUF_SIZE];
+  g_configuration->eal_initialized = 0;
+  g_configuration->helper_action = action;
+  /*
+   * Create a pipe for helper stdout back to collectd. This is necessary for
+   * logging EAL failures, as rte_eal_init() calls rte_panic().
+   */
+  if (pipe(g_configuration->helper_pipes) != 0) {
+    DEBUG("dpdkstat: Could not create helper pipe: %s", sstrerror(errno, errbuf,
+      sizeof (errbuf)));
+    return -1;
+  }
+
+  int pipe0_flags = fcntl(g_configuration->helper_pipes[0], F_GETFL, 0);
+  int pipe1_flags = fcntl(g_configuration->helper_pipes[1], F_GETFL, 0);
+  if (pipe0_flags == -1 || pipe1_flags == -1) {
+    WARNING("dpdkstat: Failed setting up pipe flags: %s", sstrerror(errno, errbuf,
+      sizeof (errbuf)));
+  }
+  int pipe0_err = fcntl(g_configuration->helper_pipes[0], F_SETFL, pipe1_flags
+    | O_NONBLOCK);
+  int pipe1_err = fcntl(g_configuration->helper_pipes[1], F_SETFL, pipe0_flags
+    | O_NONBLOCK);
+  if (pipe0_err == -1 || pipe1_err == -1) {
+    WARNING("dpdkstat: Failed setting up pipes: %s", sstrerror(errno, errbuf,
+      sizeof (errbuf)));
+  }
+
+  pid_t pid = fork();
+  if (pid > 0) {
+    close(g_configuration->helper_pipes[1]);
+    g_configuration->helper_pid = pid;
+    DEBUG("dpdkstat: helper pid %u", g_configuration->helper_pid);
+    /* Kick helper once its alive to have it start processing */
+    sem_post(&g_configuration->sema_helper_get_stats);
+  } else if (pid == 0) {
+    /* Replace stdout with a pipe to collectd. */
+    close(g_configuration->helper_pipes[0]);
+    close(STDOUT_FILENO);
+    dup2(g_configuration->helper_pipes[1], STDOUT_FILENO);
+    dpdk_helper_run();
+    exit(0);
+  } else {
+    ERROR("dpdkstat: Failed to fork helper process: %s", sstrerror(errno, errbuf,
+      sizeof (errbuf)));
+    return -1;
+  }
+  return 0;
+}
+
+/*
+ * Initialize the DPDK EAL, if this returns, EAL is successfully initialized.
+ * On failure, the EAL prints an error message, and the helper process exits.
+ */
+static int dpdk_helper_init_eal(void)
+{
+  g_configuration->helper_status = DPDK_HELPER_INITIALIZING_EAL;
+  char *argp[(g_configuration->eal_argc) + 1];
+  int i = 0;
+
+  argp[i++] = "collectd-dpdk";
+  if (strcasecmp(g_configuration->coremask, "") != 0) {
+    argp[i++] = "-c";
+    argp[i++] = g_configuration->coremask;
```

Comment:
This is assignment is not included in `eal_argc` and may overflow the allocation on the stack.